Optimize linearFV setup

GiudGiud opened the pull request:
see #31554 for before
now
```
Showing top 30 nodes out of 176
      flat  flat%   sum%        cum   cum%
     6.21s 19.73% 19.73%      6.57s 20.87%  tcmalloc::CentralFreeList::Populate
     5.76s 18.30% 38.02%      5.96s 18.93%  FaceInfo::FaceInfo
     1.75s  5.56% 43.58%      3.43s 10.90%  MooseMesh::cacheInfo
     1.57s  4.99% 48.57%      1.77s  5.62%  libMesh::DofMap::_dof_indices
     0.79s  2.51% 51.08%      8.66s 27.51%  tc_new
     0.70s  2.22% 53.30%      7.20s 22.87%  MooseMesh::buildFiniteVolumeInfo
     0.68s  2.16% 55.46%      0.68s  2.16%  hypre_BoomerAMGRelaxHybridGaussSeidel_core
     0.52s  1.65% 57.12%      1.23s  3.91%  libMesh::UnstructuredMesh::find_neighbors
     0.44s  1.40% 58.51%      0.44s  1.40%  MooseMesh::getNodeBlockIds
     0.44s  1.40% 59.91%      4.79s 15.22%  libMesh::SparsityPattern::Build::handle_vi_vj
     0.43s  1.37% 61.28%      0.49s  1.56%  PackedCache::TryGet
     0.41s  1.30% 62.58%     30.63s 97.30%  [libsasl2.2.dylib]
     0.39s  1.24% 63.82%      2.08s  6.61%  tc_delete
     0.38s  1.21% 65.03%      0.38s  1.21%  libMesh::Predicates::not_null::operator()
     0.36s  1.14% 66.17%      0.41s  1.30%  MatSetValues_SeqAIJ
     0.36s  1.14% 67.31%      0.36s  1.14%  std::__1::__cxx_atomic_load[abi:ne180100]
     0.36s  1.14% 68.46%      0.79s  2.51%  variant_filter_iterator::Pred::operator()
     0.34s  1.08% 69.54%      4.07s 12.93%  std::__1::vector::__insert_with_size[abi:ne180100]
     0.30s  0.95% 70.49%      1.03s  3.27%  std::__1::__hash_table::__emplace_unique_key_args
     0.29s  0.92% 71.41%      0.30s  0.95%  hypre_BoomerAMGBuildCoarseOperatorKT.omp_outlined.10
     0.26s  0.83% 72.24%      0.51s  1.62%  libMesh::StoredRange::reset
     0.24s  0.76% 73.00%      0.24s  0.76%  tcmalloc::Span::Span
     0.23s  0.73% 73.73%      0.32s  1.02%  tcmalloc::SLL_TryPop
     0.21s  0.67% 74.40%      0.55s  1.75%  ComputeLinearFVElementalThread::operator()
     0.19s   0.6% 75.00%      0.31s  0.98%  MooseMesh::updateActiveSemiLocalNodeRange
     0.19s   0.6% 75.60%      0.49s  1.56%  tcmalloc::ThreadCache::FreeList::Push
     0.17s  0.54% 76.14%      0.17s  0.54%  hypre_BoomerAMGBuildInterp.omp_outlined.9
     0.17s  0.54% 76.68%      0.30s  0.95%  tcmalloc::SLL_Push
     0.16s  0.51% 77.19%      0.16s  0.51%  tcmalloc::Sampler::TryRecordAllocationFast
     0.15s  0.48% 77.67%      0.25s  0.79%  std::__1::__hash_table::__do_rehash
```

I have a libMesh PR open to address 
```
     1.75s  5.56% 43.58%      3.43s 10.90%  MooseMesh::cacheInfo
```

The first commit in this PR addresses this item
```
    0.33s  0.82% 77.28%      0.36s   0.9%  std::__1::vector::__emplace_back_slow_path
```
The rest this item
```
    12.09s 30.22% 30.22%     12.86s 32.14%  FaceInfo::FaceInfo
```

GiudGiud committed:
Finite volume setup optimizations
refs #31554

GiudGiud committed:
Use a side builder for face infos

GiudGiud committed:
Prebuild some quantities for FV face info creation

GiudGiud committed:
Re-add basic constructor: easier to use for unit testing
Compute the normal at the face centroid to be able to re-use a FE at both boundary and elem

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3230752/), step Docs: sync website on be8af24 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31555/site)

This comment will be updated on new commits.

loganharbour commented on framework/include/mesh/FaceInfo.h:
```diff
@@ -36,7 +37,14 @@ class Node;
+           unsigned int side,
+           const dof_id_type id,
+           libMesh::ElemSideBuilder & side_builder,
+           libMesh::FEBase * fe);
```
```suggestion
           libMesh::FEBase & fe);
```

There's really no need to pass this as a pointer as it should never be null

loganharbour commented on framework/src/mesh/FaceInfo.C:
```diff
@@ -41,6 +41,32 @@ FaceInfo::FaceInfo(const ElemInfo * elem_info, unsigned int side, const dof_id_t
   _face_centroid = face->vertex_average();
 }
 
+FaceInfo::FaceInfo(const ElemInfo * elem_info,
```
You can reduce the duplication here by defining a common private constructor that builds everything here given a FE object and a side, and then having your main two constructors call that constructor

loganharbour reviewed (commented):
Glad to see the ElemSideBuilder is picking up some use.

@roystgnr we have this use case a ton in moose where we rely on the sides quite a bit.

In ray tracing - I have a cache of side elements that is indexed into via an extra element ID. It would be nice if we didn't have to keep up with a cache like this, but I don't see an easy way to do it without having to deal with figuring out when to sync

GiudGiud commented on framework/src/mesh/FaceInfo.C:
```diff
@@ -41,6 +41,32 @@ FaceInfo::FaceInfo(const ElemInfo * elem_info, unsigned int side, const dof_id_t
+  std::vector<Point> pts = {_face_centroid};
+  // We pass the centroid to avoid additional re-init work that fails when
+  // the same FE is used for a boundary and a volume element (of the same dimension here)
+  fe->reinit(_elem_info->elem(), _elem_side_id, libMesh::TOLERANCE, &pts);
```
this currently diffs on QUAD9
pts = face vertex average or face true centroid does not account for the difference

roystgnr replied:
My vote would be `elem->get_normal(side_index, pt)`, but yeah, the documentation on that has to be thorough; the name alone doesn't make it obvious whether `pt` should be in physical space, reference space for `elem`, or reference space for `elem->side_ptr(side_index)`.

I thought the docs for `fe->reinit(elem, side_index, tol, &pts)` got it right:
```
any points specified on the reference side element specified in the optional argument \p pts.
```
but I'd be happy with suggestions for improvements there too.

Should we add a `reference_centroid()` method to `Elem`?  `(0,0,0)` works for any tensor product element but would be subtly wrong for triangular or polygonal faces depending on what users compute there.

GiudGiud committed:
Address Logan's review

GiudGiud committed:
Use new side vertex average routine

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3452708/), step Results summary on ecb0ddb wanted to post the following:

## Framework test summary

Compared against 44696c5 in job [civet.inl.gov/job/3451277](https://civet.inl.gov/job/3451277).

No change
## Modules test summary

Compared against 44696c5 in job [civet.inl.gov/job/3451277](https://civet.inl.gov/job/3451277).

No change

GiudGiud committed:
Do not over-allocate for distributed meshes
Though we may underallocate for replicated meshes
Note: still less re-allocation than before
